### PR TITLE
Improve testing gRPC services with random ports

### PR DESCRIPTION
**Pull request by nahguam**
Currently, testing gRPC is not as seamless as HTTP when it comes to using random ports. If one injects a `@GrpcClient` when `quarkus.grpc.server.test-port=0`, it will attempt to connect to port `0`, or if one is using the Vert.x server, it will attempt to connect to the http test port, `8081`.

This change aims to improve this by making it possible to inject `@GrpcClient`s into tests that have discovered the actual port that the service is running on. It shares the approach that is used for HTTP.

### CODE REVIEW

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java [hunk 1]
```diff
@@ -1188,31 +1186,8 @@ public void handle(AsyncResult<HttpServer> event) {
...
-                                //we always set the .port property, even if we are in test mode, so this will always
-                                //reflect the current port
-                                String portPropertyName = "quarkus." + schema + ".port";
-                                String prevPortPropertyValue = System.setProperty(portPropertyName, portPropertyValue);
-                                if (!Objects.equals(prevPortPropertyValue, portPropertyValue)) {
-                                    portPropertiesToRestore.put(portPropertyName, prevPortPropertyValue);
-                                }
-                                if (launchMode == LaunchMode.TEST) {
-                                    //we also set the test-port property in a test
-                                    String testPropName = "quarkus." + schema + ".test-port";
```
Comment (nahguam) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:1202]: I have kept the original behaviour here, but this results in `quarkus.https.test-port` for Vert.x TLS tests. Is this correct, given that the formal property is `quarkus.http.test-ssl-port`?

**Comment by jbonofre** [comment]
@nahguam Thanks ! I'm doing a pass/review.

**Comment by alesj** [comment]
Where in the docs does it state if you pass 0 port to (Vert.x) gRPC server, that it randomly selects one?
(not that I'm doubting this is true ... just curious)

Plus, it would also be good/nice to fix to actual port where ever we print it out.
e.g.

```
            msg += String.format("gRPC server on %s:%d [%s]",
                    configuration.host, port, "TLS enabled: " + !configuration.plainText);
```

Otherwise it looks good + nice test coverage!

**Comment by nahguam** [comment]
> Where in the docs does it state if you pass 0 port to (Vert.x) gRPC server, that it randomly selects one? (not that I'm doubting this is true ... just curious)

To my limited knowledge, the only reference is https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/http-reference.adoc#7-listening-on-a-random-port

Should I add similar detail to the grpc doc?

> Plus, it would also be good/nice to fix to actual port where ever we print it out. e.g.
> 
> ```
>             msg += String.format("gRPC server on %s:%d [%s]",
>                     configuration.host, port, "TLS enabled: " + !configuration.plainText);
> ```

Indeed, I did have a look at this but it looked non-trivial. I'll take another look.

**Comment by nahguam** [comment]
> > Plus, it would also be good/nice to fix to actual port where ever we print it out. e.g.
> > ```
> >             msg += String.format("gRPC server on %s:%d [%s]",
> >                     configuration.host, port, "TLS enabled: " + !configuration.plainText);
> > ```
> 
> Indeed, I did have a look at this but it looked non-trivial. I'll take another look.

I've updated this output.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java [hunk 2]
```diff
@@ -226,7 +228,7 @@ private void prodStart(GrpcContainer grpcContainer, Vertx vertx, GrpcServerConfi
 
     private void postStartup(GrpcServerConfiguration configuration, GrpcBuilderProvider<?> provider, boolean test) {
         initHealthStorage();
-        int port = test ? configuration.testPort : configuration.port;
+        int port = test ? testPort(configuration) : configuration.port;
```
Comment (alesj) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java:231]: Well, this was easy. :-)
Are we sure we captured all such log outputs -- with this "new" port?

**Comment by nahguam** [comment]
> > Should I add similar detail to the grpc doc?
> 
> Are we sure this works for the gRPC as well? :-) Since I guess that is then a (pure) gRPC thing (not http / Vert.x)

The only branch that isn't tested specifically is the one that uses `GrpcBuilderProvider` so, XDS/InProcess.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java (hunk 2 above)
Comment (nahguam) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/GrpcServerRecorder.java:231]: There's an output in `buildServer` but that's before the server is started - so before the port is chosen.

**Review by cescoffier — APPROVED** [review]
LGTM!

You can retrieve the actual port from the httpServer (once bound to the socket): `server.getactualPort()).

**Review by jbonofre — APPROVED** [review]
LGTM, thanks !

**Comment by nahguam** [comment]
> > Should I add similar detail to the grpc doc?
> 
> Are we sure this works for the gRPC as well? :-) Since I guess that is then a (pure) gRPC thing (not http / Vert.x)

Do you mean when `use-separate-server=true`? If so, yes, the tests cover that.

**Comment by jbonofre** [comment]
@nahguam Clément and I approved this PR, so I think we are good to go. No need to keep rebasing, we can alway rebase when merging. @cescoffier thoughts ?

**Comment by cescoffier** [comment]
I was just waiting for 3.2 to be branched.

**Comment by jbonofre** [comment]
@cescoffier great, thank you 😄

**Comment by nahguam** [comment]
Great Thanks 👍

**Pull request merged**